src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,71 +9,347 @@
+        copy_sz =
+            cur_offset + MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE <=
+            data_sz ? MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE : data_sz - cur_offset;
+        mpi_errno =
+            MPIR_Localcopy(((char *) src_buf + cur_offset), copy_sz, MPI_BYTE,
+                           ((char *) dest_buf + cur_offset), copy_sz, MPI_BYTE);
+        MPIR_ERR_CHECK(mpi_errno);
+        num_local_copy++;
+    }
+
+    /* - If receiver has copied all chunks (num_local_copy == total_chunk), it sends fin packet to sender.
+     * - If sender and receiver both copy a part of chunks:
+     *     - when cur_chunk == total_chunk, the last chunk has been copied by the local process,
+     *        or has been assigned to the other side but may be still under processing;
+     *     - when cur_chunk == total_chunk + 1, both sides should complete. Thus the local process
+     *        sends FIN to the other side.
+     */
+    if (num_local_copy == total_chunk)
+        *fin_type = MPIDI_XPMEM_COPY_ALL_ISSUE_FIN;
+    else if (num_local_copy == 0)
+        *fin_type = MPIDI_XPMEM_WAIT_FIN;
+    else if (cur_chunk == total_chunk + 1)
+        *fin_type = MPIDI_XPMEM_COPY_MIX_ISSUE_FIN;
+    else
+        *fin_type = MPIDI_XPMEM_WAIT_FIN;
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_DO_LMT_COOP_COPY);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_enqueue_cnt_obj(MPIDI_XPMEM_cnt_free_obj_t ** queue,
+                                                         MPIDI_XPMEM_cnt_t * counter_ptr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_XPMEM_cnt_free_obj_t *cnt_obj;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_ENQUEUE_CNT_OBJ);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_ENQUEUE_CNT_OBJ);
+    MPIR_CHKPMEM_DECL(1);
+
+    MPIR_CHKPMEM_MALLOC(cnt_obj, MPIDI_XPMEM_cnt_free_obj_t *,
+                        sizeof(MPIDI_XPMEM_cnt_free_obj_t), mpi_errno, "counter obj",
+                        MPL_MEM_OTHER);
+    counter_ptr->free_flag = 0;
+    cnt_obj->counter_ptr = counter_ptr;
+    DL_APPEND(*queue, cnt_obj);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_ENQUEUE_CNT_OBJ);
+    return mpi_errno;
+  fn_fail:
+    MPIR_CHKPMEM_REAP();
+    goto fn_exit;
+}
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_collect_cnt_obj(MPIDI_XPMEM_cnt_free_obj_t ** queue)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_XPMEM_cnt_free_obj_t *cur_obj = NULL;
+    MPIDI_XPMEM_cnt_free_obj_t *next = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_COLLECT_CNT_OBJ);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_COLLECT_CNT_OBJ);
+
+    DL_FOREACH_SAFE(*queue, cur_obj, next) {
+        if (cur_obj->counter_ptr->free_flag) {
+            DL_DELETE(*queue, cur_obj);
+            MPIR_Handle_obj_free(&MPIDI_XPMEM_cnt_mem, (void *) cur_obj->counter_ptr);
+            MPL_free(cur_obj);
+        }
```

Comment:
I thought you don't want to traverse all free cnts every time?

Besides, I still concern the overhead caused by the free cnts design. I hope we can find a better solution in future. Can you please add a "TODO" comment for it?

Reply:
The new counter release logic should get rid of the above issue.